### PR TITLE
fix prefab support

### PR DESCRIPTION
**Pull request by pnwamk**
The preliminary work to get this PR ready is more-or-less done (I believe). I still need to do some "field testing" to see how these changes affect any packages out there that use Typed Racket and prefab structs.

See Typed Racket RFC [0001-prefab-structs](https://github.com/racket/typed-racket/blob/master/rfcs/text/0001-prefab-structs.md) for discussion on the overall strategy for this PR and alternatives (if we find there are issues with the current approach).

### CODE REVIEW

File: typed-racket-lib/typed-racket/types/overlap.rkt [hunk 1]
```diff
@@ -165,6 +166,19 @@
      (and t2 (Struct: _ _ _ _ _ _)))
     (or (subtype t1 t2) (subtype t2 t1)
         (parent-of? t1 t2) (parent-of? t2 t1))]
+   [((PrefabTop: key1) (or (PrefabTop: key2)
+                           (Prefab:    key2 _)))
+    #:no-order
+    (or (equal? key1 key2)
```
Comment (samth) [inline typed-racket-lib/typed-racket/types/overlap.rkt]: equal? not needed.

File: typed-racket-lib/typed-racket/types/printer.rkt [hunk 2]
```diff
@@ -171,73 +172,109 @@
...
+                #:result (cond
+                           [(not (null? pair-seq))
+                            (list (pair-seq->sym pair-seq) sexp)]
+                           [else sexp]))
+               ([pe (in-list (reverse pes))])
+       (let ([sexp (if (= 4 (length pair-seq))
+                       (list (pair-seq->sym pair-seq) sexp)
+                       sexp)])
+         (match pe
+           [(CarPE:) (values sexp (cons "a" pair-seq) (add1 depth))]
```
Comment (samth) [inline typed-racket-lib/typed-racket/types/printer.rkt]: This code should have tests (or at least examples).

File: typed-racket-lib/typed-racket/types/printer.rkt [hunk 3]
```diff
@@ -576,6 +613,9 @@
     [(Prefab: key field-types)
      `(Prefab ,(abbreviate-prefab-key key)
               ,@(map t->s field-types))]
+    [(PrefabTop: key)
+     `(PrefabTop ,(abbreviate-prefab-key key)
+                 ,(prefab-key->field-count key))]
```
Comment (samth) [inline typed-racket-lib/typed-racket/types/printer.rkt]: Make sure the RFC and how the printing works are consistent.

File: typed-racket-lib/typed-racket/types/update.rkt [hunk 4]
```diff
@@ -57,6 +57,14 @@
             [(Bottom:) -Bottom]
             [ty (let ([flds (append lhs (cons (make-fld ty acc-id #f) rhs))])
                   (make-Struct nm par flds proc poly pred))])]
+
+         ;; prefab struct ops
+         [((Prefab: key flds) (PrefabPE: path-key idx))
```
Comment (samth) [inline typed-racket-lib/typed-racket/types/update.rkt]: We're assuming we can't have paths to mutable fields so we can't get here in that case.

File: typed-racket-lib/typed-racket/utils/utils.rkt [hunk 5]
```diff
@@ -481,7 +483,7 @@ at least theoretically.
   (in-parallel (map car l) (map cdr l)))
 
 (define-sequence-syntax in-assoc
-  (λ () #'in-assoc-proc)
+  (λ () #'in-list/rest-proc)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/utils.rkt]: ?

File: typed-racket-lib/typed-racket/utils/utils.rkt [hunk 6]
```diff
@@ -511,13 +513,14 @@ at least theoretically.
 (module local-ids racket
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/utils.rkt]: This whole thing should be its own file.

File: typed-racket-test/unit-tests/prefab-tests.rkt [hunk 7]
```diff
@@ -20,71 +22,192 @@
...
+                  '(foo 2 (0 #f) #(0) bar 3 (0 #f) #(1)))
+                 (list #f #t #f #t #f))
+   (check-equal? (prefab-key->field-mutability
+                  '(foo 2 (0 #f) #(0) bar 3 (0 #f) #(1) baz 1 (0 #f) #(0)))
+                 (list #t #f #t #f #t #f))
+   ;; immutable prefab example
+   (check-eq? raw-p p0)
+   (check-equal? (pair-fst p0) 1)
+   (check-equal? (pair-snd p0) 2)
+   (check-exn exn:fail:contract:blame?
```
Comment (samth) [inline typed-racket-test/unit-tests/prefab-tests.rkt]: These should probably use something like the infrastructure in the contract test suite to check that the blame is right.

File: typed-racket-lib/typed-racket/utils/prefab-c.rkt [hunk 8]
```diff
@@ -0,0 +1,248 @@
+#lang racket/base
+
+(require racket/contract)
+
+(provide (rename-out [-prefab/c prefab/c]))
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/prefab-c.rkt]: This should probably have a comment about eagerness of checking and why it's that way (similarity with struct/c). Or maybe it should be more eager. Either way, document in RFC as well.

File: typed-racket-doc/typed-racket/scribblings/reference/types.scrbl [hunk 9]
```diff
@@ -860,21 +860,78 @@ this top type.
...
+      (person? '#s((employee person 1) Toby -1))
+      (employee? '#s((employee person 1) Toby -1))]
 }
 
+@defform[(PrefabTop key field-count)]{Describes all
+prefab types with the (implicitly quoted) prefab-key
+ @racket[key] and @racket[field-count] many fields.
+
+ For immutable prefabs this is equivalent to
+ @racket[(Prefab key Any ...)] with @racket[field-count] many
```
[the remark this answers is not in the document]
Comment (pnwamk) [reply typed-racket-doc/typed-racket/scribblings/reference/types.scrbl]: After a little more thought and review on this, it appears the use of `...` is consistent with the rest of the [TR docs](http://docs.racket-lang.org/ts-reference/type-ref.html?q=U#%28form._%28%28lib._typed-racket%2Fbase-env%2Fbase-types-extra..rkt%29._.U%29%29). I think it's best to leave it as is unless there's strong objection.

**Comment by pnwamk** [comment]
I think we're good to go---I've attempted to address the review comments with tweaks and updates to the RFC.

**Pull request merged**